### ISSUE TITLE
[BUG] create folder error

### ISSUE THREAD

**Issue by karwootang-gft** [tranquilitybase-io/tb-gcp-dac#87]
How to reproduce error:
POST below folder_async endpoint.
{
  "folder": "ABC",
  "parentFolderId": "687167067745"
}

Then query the below GET endpoint using the task_id: 188364b2-c3e5-4078-9c9f-0d4eb816172d

/folder_async/result/create/{taskid}

Response from the DAC:
{
  "detail": "The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.",
  "status": 500,
  "title": "Internal Server Error",
  "type": "about:blank"
}

Error Message in the DAC logs:

gcp-dac            | [2020-05-28 11:12:29 +0000] [20] [DEBUG] GET /api/folder_async/result/create/188364b2-c3e5-4078-9c9f-0d4eb816172d
gcp-dac            | 2020-05-28 11:12:29,159 - tb-gcp-dac - create_folder_result - INFO - CREATE FOLDER RESULT 188364b2-c3e5-4078-9c9f-0d4eb816172d
gcp-dac            | Exception on /api/folder_async/result/create/188364b2-c3e5-4078-9c9f-0d4eb816172d [GET]
gcp-dac            | Traceback (most recent call last):
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 2446, in wsgi_app
gcp-dac            |     response = self.full_dispatch_request()
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1951, in full_dispatch_request
gcp-dac            |     rv = self.handle_user_exception(e)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1820, in handle_user_exception
gcp-dac            |     reraise(exc_type, exc_value, tb)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/_compat.py", line 39, in reraise
gcp-dac            |     raise value
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1949, in full_dispatch_request
gcp-dac            |     rv = self.dispatch_request()
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1935, in dispatch_request
gcp-dac            |     return self.view_functions[rule.endpoint](**req.view_args)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/connexion/decorators/decorator.py", line 48, in wrapper
gcp-dac            |     response = function(request)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/connexion/decorators/uri_parsing.py", line 144, in wrapper
gcp-dac            |     response = function(request)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/connexion/decorators/validation.py", line 384, in wrapper
gcp-dac            |     return function(request)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/connexion/decorators/parameter.py", line 121, in wrapper
gcp-dac            |     return function(**kwargs)
gcp-dac            |   File "/app/gcpdac/folder.py", line 68, in create_folder_result
gcp-dac            |     retval = AsyncResult(taskid).get(timeout=1.0)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/celery/result.py", line 228, in get
gcp-dac            |     on_message=on_message,
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/celery/backends/asynchronous.py", line 202, in wait_for_pending
gcp-dac            |     return result.maybe_throw(callback=callback, propagate=propagate)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/celery/result.py", line 333, in maybe_throw
gcp-dac            |     self.throw(value, self._to_remote_traceback(tb))
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/celery/result.py", line 326, in throw
gcp-dac            |     self.on_ready.throw(*args, **kwargs)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/vine/promises.py", line 244, in throw
gcp-dac            |     reraise(type(exc), exc, tb)
gcp-dac            |   File "/usr/local/lib/python3.7/dist-packages/vine/five.py", line 194, in reraise
gcp-dac            |     raise value.with_traceback(tb)
gcp-dac            | TypeError: string indices must be integers

**Comment by karwootang-gft** [comment]
This error is caused by the folder_terraform script if the payload is not in the below format:

{
  "folder": {
    "folderName": "ABCDEFG",
    "parentFolderId": "940168182397"
  }
}

There should be more validation done and a task id shouldn't be returned if the required parameters are not found in the payload.

**Comment by karwootang-gft** [comment]
@sunderhill-gft - The DAC returns the folder id prefixed with "folders/" e.g. 'folders/871842960650', should the prefix folders/ be removed by the DAC before the HS receives it?

If not, should the HS remove it, and store it without the prefix and send it to the DAC without the prefix, similar to the root parent id.

Seems, like it's part of the id, and should be kept intact, not removed and added back later.

**Comment by karwootang-gft** [comment]
The mock DAC should be changed to process this format.

{
"folder": {
"folderName": "ABCDEFG",
"parentFolderId": "940168182397"
}
}

This is the format it's expecting at the moment:

{
  "folderName": "ABCDEFG",
  "parentFolderId": "940168182397"
}

**Comment by karwootang-gft** [comment]
Also, nice to have the folder details returned in the response. e.g.
{
 "payload": "****",
"status": "SUCCESS",
"folder": {
"folderName": "ABCDEFG",
"parentFolderId": "940168182397"
}

**Comment by karwootang-gft** [comment]
Each time the "create result" endpoint is called, a new task is added to Celery. 
These get queued up behind the first task (with task id) and then completed all at once once first task with task id is complete. (That's how it appears when viewing the logs). 
If a query is pending (already in the queue), then it should return immediately with PENDING status?

**Comment by karwootang-gft** [comment]
DAC delete folder endpoint is not working.

How to reproduce:
Requested a delete of the empty folder with id: 269201002740

DAC returned a task id:

{
  "taskid": "5d852d42-3ec2-4ea4-96af-fe2eca7a6624"
}

I then supported a create/results/delete using the task id and got the response:
{
  "status": "SUCCESS"
}

When I checked the cloud resource manager, the folder is still there.
Developers | 269201002740

**Comment by karwootang-gft** [comment]
If we get a request to create a folder that is "lifecycle_state" = "DELETE_REQUESTED", we should undelete the folder.


kotg@cloudshell:~ (bootstrap-29658c9f)$ gcloud resource-manager folders undelete 871842960650
Restored [<Folder
 createTime: '2020-05-28T15:11:38.058Z'
 displayName: 'Applications'
 lifecycleState: LifecycleStateValueValuesEnum(ACTIVE, 1)
 name: 'folders/871842960650'
 parent: 'folders/940168182397'>].

**Comment by karwootang-gft** [comment]
Back Office Settlements-workspace - folder name is too long, can we shorten the workspace name to Back Office Settlements-ws?

gcp-dac-worker     | WARNING:python_terraform:error: b'2020/05/29 12:41:31 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.\n  Use TF_LOG=TRACE to see Terraform\'s internal logs.\n  ----\n\nError: "Back Office Settlements-production" name must be 4 to 30 characters with lowercase and uppercase letters, numbers, hyphen, single-quote, double-quote, space, and exclamation point.\n\n  on ../environment_project/main.tf line 19, in resource "google_project" "environment_project":\n  19: resource "google_project" "environment_project" {\n\n\n\nError: "Back Office Settlements-workspace" name must be 4 to 30 characters with lowercase and uppercase letters, numbers, hyphen, single-quote, double-quote, space, and exclamation point.\n\n  on ../workspace_project/main.tf line 19, in resource "google_project" "workspace_project":\n  19: resource "google_project" "workspace_project" {\n\n\n'
gcp-dac-worker     | 2020-05-29 12:41:34,761 - tb-gcp-dac - terraform_apply - DEBUG - Terraform apply return code is 1
gcp-dac-worker     | DEBUG:tb-gcp-dac:Terraform apply return code is 1
gcp-dac-worker     | 2020-05-29 12:41:34,762 - tb-gcp-dac - terraform_apply - DEBUG - Terraform apply stdout is module.solution_folder.google_folder.folder: Refreshing state... [id=folders/832203247962]
gcp-dac-worker     | module.workspace_project.google_project.workspace_project: Refreshing state... [id=projects/workspace-2f2nuz-ksjs726s]
gcp-dac-worker     | module.environment_projects.google_project.environment_project[0]: Refreshing state... [id=projects/development-2f2nuz-ksjs726s]

**Comment by sunderhill-gft** [comment]
Original json in post request was in incorrect format so request failed. Raise issue to immediately fail if json can be checked before creating task
